### PR TITLE
Add ES module build

### LINKED ISSUE #13: ESM Module CLI Warning
Great library. But, the Angular CLI gives a warning related to CommonJS.

```
WARNING in /...REDACTED...dia-input.component.ts depends on 'image-blob-reduce'. CommonJS or AMD dependencies can cause optimization bailouts.
For more info see: https://angular.io/guide/build#configuring-commonjs-dependencies
```
I'm importing as:
```
import imageReducer from 'image-blob-reduce';
import pica from 'pica';
```

If I change the import to `import imageReducer from 'image-blob-reduce/dist/image-blob-reduce.min.js'` then my bundle size increases by 20kb (but the warning goes away).

It works either way so this is not a high priority.

### PR DESCRIPTION
**Pull request by zauni**
Fixes #13

### CODE REVIEW

**Comment by puzrin** [comment]
I understand idea, but need to discuss details.

- Why value of "main" changed? That's no related to esm.
- Probably worth replace `es` => `esm` everywhere, because `es` may me confusing. See https://github.com/rollup/rollup/issues/1917
- worth to remove `dist/*` from commit.
- Does that problem really needs fix? See https://github.com/rollup/rollup/issues/1917

**Comment by zauni** [comment]
Hi @puzrin 
thanks for your answer!

> Why value of "main" changed? That's no related to esm.

There was no "main" field in the package.json, so I added it so that the bundlers can use either "main" or "module". But I can add only "module" in this PR if you want.

> Probably worth replace es => esm everywhere, because es may me confusing. 

Yes, I can change that.

> worth to remove dist/* from commit.

Should I remove it in this PR?

> Does that problem really needs fix?

As #13 says the bundler can easier do tree shaking when the packages are in ESM style and not in CJS. And as it is quite easy to provide a ESM build with rollup it was a quick win for all consumers of this package.
And also Angular is not complaining anymore.

**Comment by puzrin** [comment]
> There was no "main" field in the package.json, so I added it so that the bundlers can use either "main" or "module". But I can add only "module" in this PR if you want.

1. Problem is not in entry name, but in value change. It should stay `index.js` to not affect development process (or until we discuss it as separate issue). Also, i prefer avoid entries with default values (if this not required for proper work after `module` added)

2. `module` seems specific for bundlers, Is it possible to use conditional exports `"exports": { "import": "./dist/...", "require": "./index.js" }` as in https://nodejs.org/api/packages.html#packages_conditional_exports?

> Should I remove it in this PR?

Yes. Autogenerated files just increase commit size and can not be verified.

**Comment by zauni** [comment]
1. Ah sorry, I missed that the index.js is already implicit the main entry of the package. Then we should not change that.

2. The `exports` field is not widely supported yet but the `module` field is supported by all bundlers like webpack/rollup/parcel. Maybe the `exports` field is something to consider in the future when all bundlers have support for it (Webpack 5 should have support, but Webpack 4 or Rollup doesn't have it at the moment see [here](https://github.com/webpack/webpack/issues/4674#issuecomment-665507118)
Or we add the `exports` field but I am not sure if we then also need the `main` field to be set. And also I am not sure if the `exports` field should point to a bundled and compiled ESM file. Maybe it would be more suitable if we would have a index.js which is written in ESM style, but not bundled/compiled.

So my proposel would be to leave the `module` field to support bundlers and not add the `exports` field at the moment.

**Comment by puzrin** [comment]
> So my proposel would be to leave the `module` field to support bundlers and not add the `exports` field at the moment.

We shound not decide for current moment only. I suggest add `exports` as standard and `module` as fallback.

**Comment by zauni** [comment]
Okay, so I added the `exports` field but in a way which is recommended here: https://nodejs.org/api/packages.html#packages_approach_1_use_an_es_module_wrapper
With this approach we have a clean index.js and a wrapper.js which is only reexporting stuff from index.js but in a module compatible way. Because this is a Node.js specific wrapper I would avoid to use the bundled ESM version from the dist folder.

But for bundlers who understand the `module` field, they get the compiled/bundled ESM version and older bundlers who understand the `browser` field get the compiled/bundled UMD version.

The only thing missing is that you have to do a `npm run build` before publishing this package so that the dist folder is generated. Should I add this to the `prepublishOnly` script?

**Comment by puzrin** [comment]
I don' like idea to fill project configs and files with tons of garbage :). What's wrong with just adding esm build and 2 fields as been proposed here: https://github.com/nodeca/image-blob-reduce/pull/16#issuecomment-726027732?

**Comment by zauni** [comment]
Thanks to years of unstandardised package.json fields, bundler evolutions and Node.js supporting CJS and ESM, we now have a lot of stuff to support :/

I just thought that the bundled/compiled version is unnecessary for Node.js users who are trying to import your package via ESM and also Node.js is warning against the so called [Dual package hazard](https://nodejs.org/api/packages.html#packages_dual_package_hazard) and the `wrapper.js` file is trying to avoid the mentioned problems.

But we can also strip it down to your version if you want. No problem with that :)

**Comment by puzrin** [comment]
Let's strip. We have a simple problem - how to reduce warnings :). It's important to keep repo clean. 

PS. AFAIK "dual package hazard" makes seance only for stateful packages.

**Comment by puzrin** [comment]
https://github.com/nodeca/image-blob-reduce/pull/16/files

- please, remove `dist/` and `.gitignore` changes from commit.

Everything else seems ok

**Comment by zauni** [comment]
I thought we want to remove generated files?

**Comment by puzrin** [comment]
> I thought we want to remove generated files?

Yeas, but with "your" files, you removed 2 addional "my" files.

Let me rephrase: do commit such way to keep `dist/` unchanged.

**Comment by puzrin** [comment]
LGTM. Thanks for patience :)

**Comment by zauni** [comment]
Thank you! 👍
